ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -516,8 +516,13 @@ public ReplicaId next() {
    * @return {@code true} if not found count in originating datacenter exceeds threshold, {@code false} otherwise.
    */
   private boolean hasFailedOnOriginatingDcNotFound() {
-    return originatingDcNotFoundFailureThreshold > 0
-        && originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold;
+    // Check that there are at least routerConfig.routerPutSuccessTarget number of replicas in leader or standby state
+    // in originating data center before failing the request. Also, in case originating data center no longer exists,
+    // avoid checking for this condition.
+    return originatingDcNotFoundFailureThreshold > 0 && (originatingDcNotFoundCount
+        >= originatingDcNotFoundFailureThreshold) && (this.originatingDcName != null
+        && getEligibleReplicas(this.originatingDcName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER)).size()
+        >= routerConfig.routerPutSuccessTarget);
```

Comment:
can you decompose this logic into more temp variables to improve the readability?